providers/implementations/keymgmt/dh_kmgmt.c

```diff
@@ -154,10 +154,30 @@ static int dh_match(const void *keydata1, const void *keydata2, int selection)
     if (!ossl_prov_is_running())
         return 0;
 
-    if ((selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0)
-        ok = ok && BN_cmp(DH_get0_pub_key(dh1), DH_get0_pub_key(dh2)) == 0;
-    if ((selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0)
-        ok = ok && BN_cmp(DH_get0_priv_key(dh1), DH_get0_priv_key(dh2)) == 0;
+    if ((selection & OSSL_KEYMGMT_SELECT_KEYPAIR) != 0) {
+        int key_checked = 0;
+
+        if ((selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0) {
```

Comment:
Nitpick: Is this `if` necessary? Doesn't `OSSL_KEYMGMT_SELECT_KEYPAIR` _always_ imply `SSL_KEYMGMT_SELECT_PUBLIC_KEY`? OK, compiler may fix it anyway...

Reply:
No, as per the decision to allow keys with just the private component.  In other words, it's perfecfly reasonable for someone to use the selector `OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS | SSL_KEYMGMT_SELECT_PRIVATE_KEY`.  It's not reasonable to do that in `EVP_PKEY_eq()`, but would be reasonable if someone wants to implement something like `EVP_PKEY_priv_eq()` with the expectation that it only works on keys that have a private key component.

Reply:
Hmm -- I'd agree if the "leading" `if` selector test above were for `OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS`. But it is `OSSL_KEYMGMT_SELECT_KEYPAIR`...

Reply:
Er, I don't follow your logic...

Reply:
Looks like I don't either :-) 

Let me try again: The new code reads:
```
if ((selection & OSSL_KEYMGMT_SELECT_KEYPAIR) != 0) {
        int key_checked = 0;

        if ((selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0) {
```

I _think_ (and try to say) it should read
```
if ((selection & OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS) != 0) {
        int key_checked = 0;

        if ((selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0) {
```
such as for me to fully understand your statement above:
> In other words, it's perfecfly reasonable for someone to use the selector OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS | SSL_KEYMGMT_SELECT_PRIVATE_KEY

Reply:
If you look further down the code, you'll see that domain parameters are checked separately.  Why should the check of the private/public key bits depend on if domain parameters were desired?

See, it's also perfectly reasonable for someone to just use the selector `OSSL_KEYMGMT_SELECT_PRIVATE_KEY`

Reply:
OK, then I misunderstood your statement regarding `OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS`. In that case the code just has a superfluous `if` statement -- which a smart compiler probably kills knowing that https://github.com/openssl/openssl/blob/398ae8231650c4bd8ddff0e5efd38233c23b1ca0/include/openssl/core_dispatch.h#L551-L552 . Sorry for the distraction. Consider my comments resolved then. Thanks again for the update (also regarding documentation).